File: leetcode/35.cpp
class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        int *p = nums.data();
        int n = nums.size();
        int l = 0, r = n - 1;
        
        while (l <= r) {
            int mid = (l+r)/2;
            
            // 这里有个 trick。可以减少一次判断。
            if (p[mid]==target){
                return mid;
            } else if (p[mid]>target) {
                r=mid-1;
            } else {
                l=mid+1;
            }
        }
        return l;
    }
};

File: leetcode/35.rs
use std::{i32, i64, u32};
use std::cell::{BorrowMutError, Cell, Ref, RefCell, RefMut};
use std::cmp::{max, min, Ordering};
use std::collections::{HashMap, HashSet, VecDeque};
use std::iter::{FromIterator, Iterator};
use std::mem::swap;
use std::rc::Rc;

impl Solution {
    fn lower_bound(a: &[i32], tg: i32) -> usize {
        let n = a.len();
        if n == 0 {
            return 0;
        }
        let (mut lb, mut ub) = (0, n - 1);
        while lb < ub {
            let mid = (lb + ub) / 2;
            match a[mid].cmp(&tg) {
                Ordering::Less => {
                    lb = mid + 1;
                }
                Ordering::Greater => {
                    ub = mid;
                }
                Ordering::Equal => {
                    return Self::lower_bound(&a[..mid], tg);
                }
            }
        }
        return {
            if a[lb] < tg {
                lb + 1
            } else {
                lb
            }
        };
    }

    pub fn search_insert(a: Vec<i32>, tg: i32) -> i32 {
        let x = Self::lower_bound(&a, tg);
        return x as i32;
    }
}
